src/python/pants/base/hash_utils.py

```diff
@@ -56,16 +49,20 @@ class CoercingEncoder(json.JSONEncoder):
   TypeError on things we might to throw at it, like a set, or a dict with tuple keys.
   """
 
-  def _maybe_encode(self, key_obj):
-    """This isn't strictly necessary, but this is the behavior I might expect."""
+  def _maybe_encode_dict_key(self, key_obj):
+    # If dict keys aren't strings, recursively encode them until they are. Checking for str here
+    # means we don't touch keys that are already strings (instead of quoting them).
     if isinstance(key_obj, str):
```

Comment:
This isinstance check should match the check in default for Iterable. Either `(bytes, str)` in both places or `str` in both places. Under python 2 the former seems to be correct, python 3 the latter suffices, I think using `(bytes, str)` covers both cases - under python 3 you'll just never get a bytes due to earlier raises.

Reply:
Yep, ran into this when testing it out in a py2 `./pants repl` without importing `unicode_literals`. Will make a change and add a test.